sphinx/util/nodes.py

```diff
@@ -33,6 +34,49 @@
 caption_ref_re = explicit_title_re  # b/w compat alias
 
 
+class NodeMatcher(object):
+    """A helper class for Node.traverse().
+
+    It checks that given node is an instance of specified node-classes and it has
+    specified node-attributes.
+
+    For example, following example searches ``reference`` node having ``refdomain``
+    and ``reftype`` attributes::
+
+        matcher = NodeMatcher(nodes.reference, refdomain='std', reftype='citation')
+        doctree.traverse(matcher)
+        # => [<citation ...>, <citation ...>, ...]
+    """
+
+    def __init__(self, *classes, **attrs):
+        # type: (nodes.Node, Any) -> None
+        self.classes = classes
+        self.attrs = attrs
+
+    def match(self, node):
+        # type: (nodes.Node) -> bool
+        try:
+            if self.classes and not isinstance(node, self.classes):
+                return False
+            else:
```

Comment:
This if-else statement might be in early-return form as following:

```
if self.classes and not isinstance(node, self.classes):
    return False

for key, value in self.attrs.items():
    if key not in node:
...
```

Early-return form reduce depth of indentation and make it easier to read source codes.

Reply:
Thanks, I think so too.